wps_shared/wps_shared/utils/s3.py

```diff
@@ -76,3 +78,65 @@ def set_s3_gdal_config():
     gdal.SetConfigOption("AWS_ACCESS_KEY_ID", config.get("OBJECT_STORE_USER_ID"))
     gdal.SetConfigOption("AWS_S3_ENDPOINT", config.get("OBJECT_STORE_SERVER"))
     gdal.SetConfigOption("AWS_VIRTUAL_HOSTING", "FALSE")
+
+
+async def apply_retention_policy_on_date_folders(
+    client: AioBaseClient,
+    bucket: str,
+    prefix: str,
+    days_to_retain: int,
+    dry_run: bool = False,
+):
+    """
+    Applies a retention policy to an S3 bucket by deleting "folders" (prefixes) named by ISO date (YYYY-MM-DD).
+
+    It deletes folders that are older than a specified number of days from today. Each folder is expected to be named
+    as a date string, such as '2024-04-01/', and located under the given prefix (e.g., 'critical_hours/2024-04-01/').
+
+    :param client: Asynchronous S3 client
+    :param bucket: The name of the S3 bucket
+    :param prefix: The prefix path within the bucket where date-named folders are located.
+    :param days_to_retain: The number of most recent days to retain. Folders older than this will be deleted.
+    :param dry_run: If True, no deletions will occur. Instead, actions will be logged to indicate what *would* be deleted.
+                    Defaults to False.
+    """
+    date_pattern = re.compile(r"/(\d{4}-\d{2}-\d{2})/$")
```

Comment:
Even though this is a pretty simple regex, I'd try to avoid them: https://softwareengineering.stackexchange.com/questions/223634/what-is-meant-by-now-you-have-two-problems

I think we can get away with a simpler check by reusing the date parser built into `datetime`, e.g.

```
def is_valid_iso_date_only(date_str):
    try:
        datetime.strptime(date_str, "%Y-%m-%d")
        return True
    except ValueError:
        return False
```